File: topics/js/greeter/dynamic_greeter.rb
class DynamicGreeter
  # is called, whenever a method is missing
  def method_missing(method_name, *arguments, &block)
    # method does not start with "greet_"
    return super unless method_name.to_s.start_with?("greet_")

    # extracts the name part from method_name
    name = method_name.to_s.split("_")[1]

    # define method that was missing
    self.class.send(:define_method, method_name) do |*args|
      "Hello, #{name.capitalize}! #{args.first || ''}"
    end

    # dynamically call the method we just defined
    send(method_name, *arguments)
  end
end

greeter = DynamicGreeter.new

puts greeter.foo

# dynamically define and call
puts greeter.greet_bob
# => "Hello, Bob! "

# dynamically define and call
puts greeter.greet_alice("How are you today?")
# => "Hello, Alice! How are you today?"

# just call
puts greeter.greet_alice("How are you today?")
# => "Hello, Alice! How are you today?"

File: topics/js/greeter/dynamic_greeter.js
class DynamicGreeter {
  constructor() {
    let greeter = this;

    // proxy object will intercept missing method calls
    return new Proxy(this, {
      get: function (target, prop, receiver) {
        // method does not start with "greet_"
        if (!prop.startsWith("greet_")) {
          return Reflect.get(...arguments);
        }

        // we have already dynamically defined the method before
        if (greeter[prop]) {
          return Reflect.get(...arguments);
        }

        // define method
        greeter[prop] = function (message = "") {
          const name = prop.split("_")[1]; // Extracts the name part from prop
          return `Hello, ${
            name.charAt(0).toUpperCase() + name.slice(1)
          }! ${message}`;
        };

        // return
        return greeter[prop];
      },
    });
  }
}

const greeter = new DynamicGreeter();

// dynamically define and call
console.log(greeter.greet_bob());
// => "Hello, Bob! "

// dynamically define and call
console.log(greeter.greet_alice("How are you today?"));
// => "Hello, Alice! How are you today?"

// just call
console.log(greeter.greet_alice("How are you today?"));
// => "Hello, Alice! How are you today?"
